Redmine#4703: new mapjson() function and support data containers in maparray()

tzz opened the pull request:
see https://cfengine.com/dev/issues/4703

* `maparray()` can now process slists, classic arrays, and data containers
* new function `mapjson()` exactly like `maparray()` but the output is parsed as JSON into a data container
* new `this.k2` variable, for when we iterate through a second-level slist or container

Documentation and examples will follow upon merge.

The fixes for https://cfengine.com/dev/issues/4704 (marked as TODO in the acceptance test) will follow upon merge.

tzz commented:
ping @estenberg I can rebase this PR if you think it has a chance of getting merged

estenberg commented:
This looks useful as a feature.
Only a bit concerned about the risk, especially with respect to changing maparray().

But I'm not the best to judge that, perhaps something for @kacfengine

kacfengine commented:
This is a pretty big pull request in an area which I am, admittedly, not *that* familiar. I'm a little confused about how the ``k2`` variable fits into all this. In your test case I can't see that it ever expands to anything, but it's also a complex beast, so maybe I'm missing something.

The ``maparray`` implementation does change significantly, but the external behavior is the same, if I understand correctly, and I'd definitely prefer to keep the two sharing code like you're doing here. It seems quite well tested as well, so I'm not so worried about that part.

tzz commented:
I rebased the PR but can't test it on Mac OS X currently, so more to come...

`k2` is the secondary iteration key.  It's used when `maparray()` is going through an slist under a key, and when `mapdata()` is going through a two-level map.

tzz commented:
The `mapdata()` function works exactly like `maparray()` as far as parameters: they both will take an array name or a data container and iterate over the keys.  For each key-value pair they will then apply the template.  But there are two enhancements:

* `k2` is the secondary iteration key and allows iteration over multiple values when the value is a "classic" CFE array entry of type `slist` (before, `maparray()` skipped `slist`s).  Similarly, when iterating over a two-level data container, `k2` will be the key at the second level.
* `mapdata()` doesn't just map to a slist.  It parses every element of that slist as JSON and returns the resulting array.  So `{ "key": "$(this.k)", "value": "$(this.v)" }` will extract the key and value, for example.  And `["$(this.v)" ]` will result in an array of arrays, each with just one element, the value.

tzz commented:
Here's a practical example of `k2`:

```
      "load6" data => parsejson('
{
  "a": { "a1": "V1", "a2": "V2" }, 
  "b": { "b1": "V3", "b2": "V4" }
}');

      "mapped" data => mapdata('{ "key": "$(this.k)", "key2": "$(this.k2)", "value": "$(this.v)" }', 'load6');
```
This will result in `mapped` being this data container:

```json
[
  {"key":"a","key2":"a1","value":"V1"},
  {"key":"a","key2":"a2","value":"V2"},
  {"key":"b","key2":"b1","value":"V3"},
  {"key":"b","key2":"b2","value":"V4"}
]
```

tzz commented:
I added a followup commit that, when iterating over slists or second-level JSON arrays, sets `k2` to the offset.  So for example:

```
      "load2[mykey]" slist => { "a", "b", "c" };
      "jsonspec" string => '{ "key": "$(this.k)", "key2": "$(this.k2)", "value": "$(this.v)" }';
      "mapdata_json2" data => mapdata("json", $(jsonspec), load2);
```

produces this in `mapdata_json2`:

```json
[
  {"key":"mykey","key2":"0","value":"a"},
  {"key":"mykey","key2":"1","value":"b"},
  {"key":"mykey","key2":"2","value":"c"}
]
```

If this is acceptable, I'll update the tests and provide docs.

kacfengine commented:
Thanks, that explains it better. I'm thinking that ``k2`` is just the first of an infinite number of possible k indexes though, depending on depth. Maybe it's better to use the array notation to indicate that, so ``k[0]`` and ``k[1]`` instead of ``k`` and ``k2`` (or it could start at 1).

tzz commented:
@kacfengine the `this.k` notation predates this PR and was patterned after `this.v` which was in the original `maplist()` function from eons ago.  I think changing that would break people's policies.  The other concern with changing those is that, if you look in the change of `vars.c`, the new variable names have to be explicitly listed so they don't get expanded.

Finally, yes, we could go down to `k3` and so on, but I think the code will get progressively more complex and less useful and I can't think of any uses for it.  I tried :)

tzz commented:
ping?

tzz commented:
ping?  I have rebased this 4 times now!

kacfengine commented:
Sorry, my queue is long these days. I'm taking a look now.

ediosyncratic commented on libpromises/evalfunction.c:
```diff
@@ -2367,9 +2367,24 @@ static FnCallResult FnCallLsDir(ARG_UNUSED EvalContext *ctx, ARG_UNUSED const Po
-static FnCallResult FnCallMapArray(EvalContext *ctx,
-                                   ARG_UNUSED const Policy *policy,
-                                   const FnCall *fp, const Rlist *finalargs)
+bool EvalContextVariablePutSpecialEscaped(EvalContext *ctx, SpecialScope scope, const char *lval, const void *value, DataType type, const char *tags, bool escape)
```
Split long line.

kacfengine commented:
I think it's ok to not allow deeper indexes, I agree that there's no reason to increase code complexity for that, and the use cases for that seem quite scarce. We can simply document that you can't go deeper than ``k2``.

But I'd still change it to be called ``k[1]``, indicating that ``k[0] == k`` (I would make those synonymous if it's not too much trouble). Reasons:

1. We might like to expand into ``k3`` later, so it's more future proof.
1. Using arrays is the standard way to index variables in CFEngine. ``k2`` looks more like a hack.

tzz commented:
@kacfengine  I don't feel confident enough to implement it as you suggest, so perhaps it's best you take over and rebase it at this point.  Sorry I can't dedicate more time to this, it's taken **many** hours already.

kacfengine commented:
I don't have a lot of time to spend on this either, unfortunately. I tried the very simple act of replacing ``k2`` with ``k[1]`` everywhere, which resulted in the test failing. However, it seems that all the values are the same, it's just the array order which has changed. Is this significant for the test?

This is the output:

<pre>
R: maparray 5+1: actual { "key = mykey", "key = lastkey!", "key = anotherkey" } != 5+1 expected { "key = lastkey!", "key = anotherkey", "key = mykey" }
R: maparray 5+2: actual { "key = mykey, value = myvalue", "key = lastkey!, value = o\"ne", "key = lastkey!, value = two", "key = lastkey!, value = three", "key = anotherkey, value = anothervalue" } != 5+2 expected { "key = lastkey!, value = o\"ne", "key = lastkey!, value = two", "key = lastkey!, value = three", "key = anotherkey, value = anothervalue", "key = mykey, value = myvalue" }
R: maparray 5+3: actual { "key = mykey, key2 = 0, value = myvalue", "key = lastkey!, key2 = 0, value = o\"ne", "key = lastkey!, key2 = 1, value = two", "key = lastkey!, key2 = 2, value = three", "key = anotherkey, key2 = $(this.k[1]), value = anothervalue" } != 5+3 expected { "key = lastkey!, key2 = $(this.k[1]), value = o\"ne", "key = lastkey!, key2 = $(this.k[1]), value = two", "key = lastkey!, key2 = $(this.k[1]), value = three", "key = anotherkey, key2 = $(this.k[1]), value = anothervalue", "key = mykey, key2 = $(this.k[1]), value = myvalue" }
R: maparray 5+4: actual { "xvalue should be myvalue", "xvalue should be lastvalue", "xvalue should be anothervalue" } != 5+4 expected { "xvalue should be lastvalue", "xvalue should be anothervalue", "xvalue should be myvalue" }
R: mapdata_json 5+2: actual [{"key":"mykey","value":"myvalue"},{"key":"lastkey!","value":"o\"ne"},{"key":"lastkey!","value":"two"},{"key":"lastkey!","value":"three"},{"key":"anotherkey","value":"anothervalue"}] != 5+2 jexpected [{"key":"lastkey!","value":"o\"ne"},{"key":"lastkey!","value":"two"},{"key":"lastkey!","value":"three"},{"key":"anotherkey","value":"anothervalue"},{"key":"mykey","value":"myvalue"}]
R: mapdata_json 5+3: actual [["mykey","myvalue"],["lastkey!","o\"ne"],["lastkey!","two"],["lastkey!","three"],["anotherkey","anothervalue"]] != 5+3 jexpected [["lastkey!","o\"ne"],["lastkey!","two"],["lastkey!","three"],["anotherkey","anothervalue"],["mykey","myvalue"]]
R: mapdata_json 5+4: actual [{"key":"mykey","key2":"0","value":"myvalue"},{"key":"lastkey!","key2":"0","value":"o\"ne"},{"key":"lastkey!","key2":"1","value":"two"},{"key":"lastkey!","key2":"2","value":"three"},{"key":"anotherkey","key2":"$(this.k[1])","value":"anothervalue"}] != 5+4 jexpected [{"key":"lastkey!","key2":"$(this.k[1])","value":"o\"ne"},{"key":"lastkey!","key2":"$(this.k[1])","value":"two"},{"key":"lastkey!","key2":"$(this.k[1])","value":"three"},{"key":"anotherkey","key2":"$(this.k[1])","value":"anothervalue"},{"key":"mykey","key2":"$(this.k[1])","value":"myvalue"}]
R: mapdata_none 5+1: actual ["key = mykey","key = lastkey!","key = anotherkey"] != 5+1 jnexpected ["key = lastkey!","key = anotherkey","key = mykey"]
R: mapdata_none 5+2: actual ["key = mykey, value = myvalue","key = lastkey!, value = o\"ne","key = lastkey!, value = two","key = lastkey!, value = three","key = anotherkey, value = anothervalue"] != 5+2 jnexpected ["key = lastkey!, value = o\"ne","key = lastkey!, value = two","key = lastkey!, value = three","key = anotherkey, value = anothervalue","key = mykey, value = myvalue"]
R: mapdata_none 5+3: actual ["key = mykey, key2 = 0, value = myvalue","key = lastkey!, key2 = 0, value = o\"ne","key = lastkey!, key2 = 1, value = two","key = lastkey!, key2 = 2, value = three","key = anotherkey, key2 = $(this.k[1]), value = anothervalue"] != 5+3 jnexpected ["key = lastkey!, key2 = $(this.k[1]), value = o\"ne","key = lastkey!, key2 = $(this.k[1]), value = two","key = lastkey!, key2 = $(this.k[1]), value = three","key = anotherkey, key2 = $(this.k[1]), value = anothervalue","key = mykey, key2 = $(this.k[1]), value = myvalue"]
R: mapdata_none 5+4: actual ["xvalue should be myvalue","xvalue should be lastvalue","xvalue should be anothervalue"] != 5+4 jnexpected ["xvalue should be lastvalue","xvalue should be anothervalue","xvalue should be myvalue"]
R: "jnexpected_5_1" string => '["key = mykey","key = lastkey!","key = anotherkey"]';
R: "jnexpected_5_2" string => '["key = mykey, value = myvalue","key = lastkey!, value = o\"ne","key = lastkey!, value = two","key = lastkey!, value = three","key = anotherkey, value = anothervalue"]';
R: "jnexpected_5_3" string => '["key = mykey, key2 = 0, value = myvalue","key = lastkey!, key2 = 0, value = o\"ne","key = lastkey!, key2 = 1, value = two","key = lastkey!, key2 = 2, value = three","key = anotherkey, key2 = $(this.k[1]), value = anothervalue"]';
R: "jnexpected_5_4" string => '["xvalue should be myvalue","xvalue should be lastvalue","xvalue should be anothervalue"]';
R: /root/code/core/tests/acceptance/./01_vars/02_functions/maparray_mapdata.cf FAIL
</pre>

tzz commented:
Oh, I see the problem.  We need canonical JSON output (keys must be sorted).  Otherwise this will keep happening.  Fortunately the fix is simple.  See https://github.com/cfengine/core/pull/2172 and when that's in, we can proceed here.

kacfengine commented:
Canonical JSON has been merged now. Wanna retry this with ``k2`` replaced with ``k[1]`` everywhere?

tzz commented:
OK, rebased, changed to use `k[1]` and tested heavily.  The new `dcs_check_state()` bundle could only work in 3.7 because of several new features like `mergedata()` wrapping, full JSON dumps in Mustache, and the `bundlestate()` function.  It's pretty wonderful. @nickanderson should take a look, it could make it **REALLY** easy to write tests that verify pure-data functions and more.

tzz commented:
@nickanderson apparently `mapdata()`, in reimplementing `maparray()`, breaks test `/01_vars/02_functions/maparray_multi_index.cf` which you just added recently.  Damn.  I will look at it.

tzz commented:
Well, this was fun... we now have multilevel CFEngine "classic" arrays supported in `maparray()`, `mapdata()`, `mergedata()`, and the folding functions (everywhere we make a JSON container from the array). I added a built-in count in every ref to remember how deep it was nested, so later we can start at that depth when making a JSON container from it.  So that things like `mergedata("myarray[x]")` will give you the stuff under `x`, not `{ "x": ...the stuff under x ... }`.  I hope this is a useful bridge between "classic" CFEngine policy and JSON-based policy.

All the acceptance tests pass and I believe I've done it correctly, but it could use a review.  I feel all right about it, because our acceptance tests cover a lot of the corner cases.

kacfengine commented:
> So that things like mergedata("myarray[x]") will give you the stuff under x, not { "x": ...the stuff under x ... }. I hope this is a useful bridge between "classic" CFEngine policy and JSON-based policy.

Sounds pretty useful to me, and more intuitive.

Tests will indeed be our friend here. As mentioned earlier, Sigurd was the expert in this area, not me, so his review would be more comprehensive. But the tests indeed cover a lot of cases, and it looks good to my eyes.

kacfengine commented:
trigger build

kacfengine commented:
The build is green on this one. Can be merged as soon as there are some docs.

tzz commented:
docs in https://github.com/cfengine/documentation/pull/1023

After this is merged I'll start using the new state-based acceptance tests.

tzz commented:
`mapdata()` and `maparray()` examples added as well

kacfengine commented:
trigger build

kacfengine commented:
Thanks, nice work!

tzz commented:
I think this sets the record for my longest-lived PR to get accepted :)

kacfengine commented:
Hehe yes.

If you want to beat it, start the pull request "Bump version to 3.10", and then wait for that to happen! :-)
